runtime/mount/agent.py

```diff
@@ -0,0 +1,684 @@
+import socket
+import sys
+import traceback
+import uuid
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import Literal
+from textwrap import dedent
+
+sandbox_root = os.environ.get("LATCH_SANDBOX_ROOT")
+if sandbox_root:
+    import pathlib
+    original_path_new = pathlib.Path.__new__
+
+    def patched_path_new(cls, *args, **kwargs):
+        if args and args[0] == "/root/.latch":
+            return original_path_new(cls, sandbox_root, *args[1:], **kwargs)
+        return original_path_new(cls, *args, **kwargs)
+
+    pathlib.Path.__new__ = patched_path_new
+
+AGENT_DEBUG = os.environ.get("AGENT_DEBUG") == "1"
+
+from agents import Agent, Runner, SQLiteSession, function_tool
+from agents.items import ReasoningItem
+from agents.model_settings import ModelSettings
+from agents.result import RunResult
+from instructions import construct_instructions
+from lplots import _inject
+from openai.types.shared.reasoning import Reasoning
+from pydantic import BaseModel
+from socketio_thread import SocketIoThread
+
+
+class Mode(Enum):
+    planning = "planning"
+    executing = "executing"
+    debugging = "debugging"
+
+
+class PlanItem(BaseModel):
+    id: str
+    description: str
+    status: Literal["todo", "in_progress", "done"]
+
+
+class PlanDiff(BaseModel):
+    action: Literal["add", "update", "complete"]
+    id: str
+    description: str
+
+
+class NotebookResponse(BaseModel):
+    plan: list[PlanItem]
+    plan_diff: list[PlanDiff]
+    summary: list[str] | None = None
+    questions: list[str] | None = None
+
+
+from typing_extensions import TypedDict
+
+class PlanItemPayload(TypedDict):
+    id: str
+    description: str
+    status: Literal["todo", "in_progress", "done"]
+
+
+class PlanDiffPayload(TypedDict):
+    action: Literal["add", "update", "complete"]
+    id: str
+    description: str
+
+
+@dataclass
+class AgentHarness:
+    conn: SocketIoThread
+    initialized: bool = False
+    api_key: str | None = None
+    agent: Agent | None = None
+    session: SQLiteSession | None = None
+    mode: Mode = Mode.planning
+    pending_operations: dict[str, asyncio.Future] = field(default_factory=dict)
+    executing_cells: set[str] = field(default_factory=set)
+    tools: list = field(default_factory=list)
+    active_tasks: set[asyncio.Task] = field(default_factory=set)
+    error_fixes_in_progress: set[str] = field(default_factory=set)
+    operation_counter: int = 0
+
+    mode_config: dict[Mode, tuple[str, str]] = field(default_factory=lambda: {
+        Mode.planning: ("gpt-5", "medium"),
+        Mode.executing: ("gpt-5", "low"),
+        Mode.debugging: ("gpt-5", "medium"),
+    })
+
+    async def send(self, msg: dict[str, object]) -> None:
+        await self.conn.send(msg)
+
+    async def atomic_operation(self, action: str, params: dict) -> dict:
+        self.operation_counter += 1
+
+        if self.mode == Mode.planning and action in {"create_cell", "edit_cell", "run_cell", "delete_cell"}:
```

Comment:
seems like you should remove the self.mode == Mode.planning and just make set_mode idempotent? Idk might be missing something but seems like if the previous mode was debugging it should go back to executing.